File: communication/src/main/java/net/tomofiles/skysign/communication/domain/communication/Communication.java
package net.tomofiles.skysign.communication.domain.communication;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import net.tomofiles.skysign.communication.event.EmptyPublisher;
import net.tomofiles.skysign.communication.event.Publisher;

@EqualsAndHashCode(of = {"id"})
@ToString
public class Communication {
    @Getter
    private final CommunicationId id;

    private final Generator generator;

    @Getter(value = AccessLevel.PACKAGE)
    @Setter(value = AccessLevel.PACKAGE)
    private Telemetry telemetry = null;

    @Getter(value = AccessLevel.PACKAGE)
    private final List<Command> commands;

    @Getter(value = AccessLevel.PACKAGE)
    private final List<UploadMission> uploadMissions;

    @Setter
    private Publisher publisher = new EmptyPublisher();
    
    Communication(CommunicationId id, Generator generator) {
        this.id = id;
        this.commands = new ArrayList<>();
        this.uploadMissions = new ArrayList<>();

        this.generator = generator;
    }

    public void pushTelemetry(TelemetrySnapshot snapshot) {
        this.telemetry = Telemetry.newInstance()
                .setPosition(
                        snapshot.getLatitude(),
                        snapshot.getLongitude(),
                        snapshot.getAltitude(),
                        snapshot.getRelativeAltitude(),
                        snapshot.getSpeed())
                .setArmed(snapshot.isArmed())
                .setFlightMode(snapshot.getFlightMode())
                .setOrientation(
                        snapshot.getX(),
                        snapshot.getY(),
                        snapshot.getZ(),
                        snapshot.getW());

        this.publisher.publish(
            new TelemetryUpdatedEvent(
                this.id,
                snapshot
            )
        );
    }

    public TelemetrySnapshot pullTelemetry() {
        return new TelemetrySnapshot(
                this.telemetry.getPosition().getLatitude(),
                this.telemetry.getPosition().getLongitude(),
                this.telemetry.getPosition().getAltitude(),
                this.telemetry.getPosition().getRelativeAltitude(),
                this.telemetry.getSpeed(),
                this.telemetry.isArmed(),
                this.telemetry.getFlightMode(),
                this.telemetry.getOrientation().getX(),
                this.telemetry.getOrientation().getY(),
                this.telemetry.getOrientation().getZ(),
                this.telemetry.getOrientation().getW()
        );
    }

    public List<CommandId> getCommandIds() {
        return this.commands.stream()
                .sorted(Comparator.comparing(Command::getTime))
                .map(Command::getId)
                .collect(Collectors.toList());
    }

    public CommandId pushCommand(CommandType commandType) {
        if (ArmCommandPushPolicy.isFollow(commandType, this)) {
            this.pushCommandDo(CommandType.ARM);
        }
        return this.pushCommandDo(commandType);
    }

    public CommandId pushUploadMission(MissionId missionId) {
        CommandId id = this.pushCommandDo(CommandType.UPLOAD);
        this.uploadMissions.add(new UploadMission(id, missionId));
        return id;
    }

    CommandId pushCommandDo(CommandType commandType) {
        CommandId id = this.generator.newCommandId();
        LocalDateTime time = this.generator.newTime();
        this.commands.add(new Command(id, commandType, time));
        return id;
    }

    public CommandType pullCommandById(CommandId id) {
        Command command = this.commands.stream()
                .filter(Command.empty(id)::equals)
                .findAny()
                .orElse(null);
        if (command == null) {
            return null;
        }
        this.commands.remove(command);
        return command.getType();
    }

    public MissionId pullUploadMissionById(CommandId id) {
        UploadMission uploadMission = this.uploadMissions.stream()
                .filter(UploadMission.empty(id)::equals)
                .findAny()
                .orElse(null);
        if (uploadMission == null) {
            return null;
        }
        this.uploadMissions.remove(uploadMission);
        return uploadMission.getMissionId();
    }
}

File: remote-communication/pkg/communication/domain/communication/communication.go
package communication

import (
	"errors"
	"remote-communication/pkg/common/domain/event"
	"sort"
	"time"
)

// ID .
type ID string

var (
	// ErrCannotPullCommand .
	ErrCannotPullCommand = errors.New("cannot pull command by id")
	// ErrCannotPullUploadMission .
	ErrCannotPullUploadMission = errors.New("cannot pull upload mission by id")
)

// Communication .
type Communication struct {
	id             ID
	telemetry      *Telemetry
	commands       []*Command
	uploadMissions []*UploadMission
	gen            Generator
	pub            event.Publisher
}

// SetPublisher .
func (c *Communication) SetPublisher(pub event.Publisher) {
	c.pub = pub
}

// GetID .
func (c *Communication) GetID() ID {
	return c.id
}

// PushTelemetry .
func (c *Communication) PushTelemetry(snapshot TelemetrySnapshot) {
	c.telemetry = NewTelemetryBySnapshot(snapshot)

	if c.pub != nil {
		c.pub.Publish(TelemetryUpdatedEvent{
			CommunicationID: c.id,
			Telemetry:       snapshot,
		})
	}
}

// PullTelemetry .
func (c *Communication) PullTelemetry() TelemetrySnapshot {
	return c.telemetry.GetSnapshot()
}

// GetCommandIDs .
func (c *Communication) GetCommandIDs() []CommandID {
	var commandIDs []CommandID
	commands := append([]*Command{}, c.commands...)
	sort.Slice(commands, func(i, j int) bool {
		return commands[i].time.Before(commands[j].time)
	})
	for _, command := range commands {
		commandIDs = append(commandIDs, command.id)
	}
	return commandIDs
}

// PushCommand .
func (c *Communication) PushCommand(cType CommandType) CommandID {
	if IsFollowArmCommandPushPolicy(cType, c) {
		c.pushCommandDo(CommandTypeARM)
	}
	return c.pushCommandDo(cType)
}

// PushUploadMission .
func (c *Communication) PushUploadMission(missionID MissionID) CommandID {
	id := c.pushCommandDo(CommandTypeUPLOAD)
	c.uploadMissions = append(c.uploadMissions, NewUploadMission(id, missionID))
	return id
}

func (c *Communication) pushCommandDo(cType CommandType) CommandID {
	id := c.gen.NewCommandID()
	time := c.gen.NewTime()
	c.commands = append(c.commands, NewCommand(id, cType, time))
	return id
}

// PullCommandById .
func (c *Communication) PullCommandByID(commandID CommandID) (CommandType, error) {
	var command *Command
	var commands []*Command
	for _, cmd := range c.commands {
		if cmd.id == commandID {
			command = cmd
		} else {
			commands = append(commands, cmd)
		}
	}
	if command == nil {
		return "", ErrCannotPullCommand
	}
	c.commands = commands
	return command.cType, nil
}

// PullUploadMissionByID .
func (c *Communication) PullUploadMissionByID(commandID CommandID) (MissionID, error) {
	var uploadMission *UploadMission
	var uploadMissions []*UploadMission
	for _, um := range c.uploadMissions {
		if um.commandID == commandID {
			uploadMission = um
		} else {
			uploadMissions = append(uploadMissions, um)
		}
	}
	if uploadMission == nil {
		return "", ErrCannotPullUploadMission
	}
	c.uploadMissions = uploadMissions
	return uploadMission.missionID, nil
}

// Generator .
type Generator interface {
	NewCommandID() CommandID
	NewTime() time.Time
}
